conn.go

```diff
@@ -565,21 +571,36 @@ func (c *Conn) Read(buf []byte, offset uint64, flags CommandFlags) (n int, err e
 			if err := binary.Read(c.conn, binary.BigEndian, &hole); err != nil {
 				return n, fmt.Errorf("read hole offset from chunk: %w", err)
 			}
+
+			got := span.Span[uint64]{
+				Start: hole.Offset,
+				End:   hole.Offset + uint64(hole.Length),
+			}
+
+			if err := got.Check(); err != nil || !wantedBlocks.Contains(got) {
+				return n, newOutOfRangeError(wantedBlocks, got, "hole reply")
+			}
+
 			n += int(hole.Length)
 		case nbdproto.REPLY_TYPE_OFFSET_DATA:
 			var absoluteOffset uint64
 			if err := binary.Read(c.conn, binary.BigEndian, &absoluteOffset); err != nil {
 				return n, fmt.Errorf("read data offset from chunk: %w", err)
 			}
 
-			normalizedOffset := absoluteOffset - offset
+			datalen := uint64(hdr.structured.Length) - uint64(unsafe.Sizeof(absoluteOffset))
 
-			datalen := int(hdr.structured.Length) - int(unsafe.Sizeof(absoluteOffset))
+			got := span.Span[uint64]{
+				Start: absoluteOffset,
+				End:   absoluteOffset + datalen,
+			}
 
-			if int(normalizedOffset)+int(datalen) > len(buf) {
-				return n, errors.New("server chunk is too large for given buf")
+			if err := got.Check(); err != nil || !wantedBlocks.Contains(got) {
```

Comment:
I read span first, it looks like this is only usage of Contains(), and the Check() is done here. Forcing all callers to check that spans are valid and making the methods assume they are is a classic optimization, makes sense if documented, but it might be prone to misuse.